File: Topic Wise/Hashing/groupAnagrams.py
# Group Anagrams LC 49

strs=["eat", "tea", "tan", "ate", "nat", "bat"]

def groupAnagrams(strs):
    res={}
    for s in strs:
        # print(s)
        key=''.join(sorted(s))
        # print(key)
        # print(list(res))
        if key in res:
            res[key].append(s)
            # print(res)
        else:
            res[key]=[s]
            # print(res)

    return list(res.values())

print(groupAnagrams(strs))    

File: Topic Wise/Hashing/groupAnagrams.java
import java.util.*;
class groupAnagrams {
    public static void main(String[] args) {
        String[] words = { "eat", "tea", "tan", "ate", "nat", "bat" };
        System.out.println(groupanagrams(words));
    }

    public static List<List<String>> groupanagrams(String[] words) {
        List<List<String>> result = new ArrayList<>();
        Map<String, List<String>> map = new HashMap<>();
        for (String word : words) {
            char[] charArray = word.toCharArray();
            Arrays.sort(charArray);
            String sortedWord = new String(charArray);
            if (!map.containsKey(sortedWord)) {
                map.put(sortedWord, new ArrayList<>());
            }
            map.get(sortedWord).add(word);
        }
        result.addAll(map.values());
        return result;
    }
}